### PR TITLE
wrote README.md

### PR DESCRIPTION
**Pull request by RohanBhirangi**
Added basic README.md
Describes in-toto and gives a simple summary of the working.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,29 @@
 # in-toto
 in-toto is a series of scripts to protect the supply chain integrity.
+
+# About in-toto
+A software supply chain is the series of steps performed when writing, testing, packaging, and distributing software. A typical software supply chain is composed of multiple steps "chained" together that transform (e.g., compilation) or verify the state (e.g., linting) of the project in order to drive it to a final product.
+
+In simple words, in-toto guarantees that the end-user (or owner) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (participants) have perfomed the specified tasks and there hasn't been any malicious changes in the files.
```
Comment (SantiagoTorres) [inline README.md]: I think this could be split in a couple of shorter, simpler sentences.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,29 @@
...
+- Alice speficies the layout in JSON format with Bob and Carl's public keys and signs it with her private key.
+- Bob writes a python script and in-toto hashes the product and stores it to a piece of link metadata. The metadata is signed with Bob's private key. Bob sends the python script to Carl for packaging.
+- Carl packages the script into a tarball and in-toto stores the hash to the metadata.
+- The client (end-user) can now verify the final product which contains the layout, tarball, the both link metadata. The in-toto-verify command basically checks whether the layout hasn't expried and if the steps were performed according to the layout definitions.
+
+# Using in-toto
+We have an open-source implementation, which is still under heavy development.
+You can clone our repository and run in-toto commands (run, verify, etc.) from command line.
+
+# Contributing to in-toto
```
Comment (SantiagoTorres) [inline README.md]: Thanks for adding this section! it definitely goes in, but it's weird to have it just empty with nothing in it right?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,29 @@
 # in-toto
 in-toto is a series of scripts to protect the supply chain integrity.
+
+# About in-toto
+A software supply chain is the series of steps performed when writing, testing, packaging, and distributing software. A typical software supply chain is composed of multiple steps "chained" together that transform (e.g., compilation) or verify the state (e.g., linting) of the project in order to drive it to a final product.
+
+In simple words, in-toto guarantees that the end-user (or owner) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (participants) have perfomed the specified tasks and there hasn't been any malicious changes in the files.
+
+So what in-toto basically does is that it asks for a project layout (JSON) stating the layout's expiry, functionary keys and steps involved. The project owner is tasked with layout specification. The layout also specifies the functionaries and the tasks they are supposed to perform.
```
Comment (SantiagoTorres) [inline README.md]: I'd personally drop the (JSON) part, considering that maybe we'll move on to another standard.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,29 @@
...
+The demo basically outlines three users viz., Alice (project owner), Bob (functionary) and Carl (functionary) and how in-toto helps to specifiy a project layout and verify that the layout has been followed in a correct manner.
+
+- Alice speficies the layout in JSON format with Bob and Carl's public keys and signs it with her private key.
+- Bob writes a python script and in-toto hashes the product and stores it to a piece of link metadata. The metadata is signed with Bob's private key. Bob sends the python script to Carl for packaging.
+- Carl packages the script into a tarball and in-toto stores the hash to the metadata.
+- The client (end-user) can now verify the final product which contains the layout, tarball, the both link metadata. The in-toto-verify command basically checks whether the layout hasn't expried and if the steps were performed according to the layout definitions.
+
+# Using in-toto
+We have an open-source implementation, which is still under heavy development.
+You can clone our repository and run in-toto commands (run, verify, etc.) from command line.
```
Comment (SantiagoTorres) [inline README.md]: This reads a little bit odd, given that's the README of the OSS implementation itself :P. Maybe we could say something else here?

File: README.md [hunk 5]
```diff
@@ -1,2 +1,29 @@
...
+
+# About in-toto
+A software supply chain is the series of steps performed when writing, testing, packaging, and distributing software. A typical software supply chain is composed of multiple steps "chained" together that transform (e.g., compilation) or verify the state (e.g., linting) of the project in order to drive it to a final product.
+
+In simple words, in-toto guarantees that the end-user (or owner) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (participants) have perfomed the specified tasks and there hasn't been any malicious changes in the files.
+
+So what in-toto basically does is that it asks for a project layout (JSON) stating the layout's expiry, functionary keys and steps involved. The project owner is tasked with layout specification. The layout also specifies the functionaries and the tasks they are supposed to perform.
+After each functionary performs its task, a link metadata is generated.
+This metadata forms a part of the final product where it is used by the client to verify that there hasn't been any changes in the files between different steps.s
```
Comment (SantiagoTorres) [inline README.md]: Is this incomplete? there's a spureous s at the end.

**Review by SantiagoTorres — CHANGES_REQUESTED** [review]
Thanks for taking care of this! Overall, I think it's in a good direction, but I left some small comments that you can use to tweak this PR.

File: README.md [hunk 6]
```diff
@@ -1,2 +1,49 @@
...
+
+# About in-toto
+in-toto guarantees that the end-user (or owner) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (participants) have perfomed the specified tasks and there haven't been any malicious changes in the files.
+
+in-toto requires a project layout that specifies the functionaries and the tasks they are supposed to perform.
+After each functionary performs its task, a link metadata is generated.
+This metadata is used to verify the intermediate and final products with the project layout.
+
+# in-toto demo
+You can try in-toto by running the demo application at [https://github.com/in-toto/in-toto/tree/develop/demo].
```
Comment (lukpueh) [inline README.md]: Please use GitHub's markdown link markup, e.g. `[demo application](https://github.com/in-toto/in-toto/tree/develop/demo)`

File: README.md [hunk 7]
```diff
@@ -1,2 +1,49 @@
...
+The demo basically outlines three users viz., Alice (project owner), Bob (functionary) and Carl (functionary) and how in-toto helps to specifiy a project layout and verify that the layout has been followed in a correct manner.
+
+- Alice speficies the layout with Bob and Carl's public keys and signs it with her private key.
+- Bob writes a python script and in-toto hashes the product and stores it to a piece of link metadata. The metadata is signed with Bob's private key. Bob sends the python script to Carl for packaging.
+- Carl packages the script into a tarball and in-toto stores the hash to the metadata.
+- The client (end-user) can now verify the final product which contains the layout, tarball, the both link metadata. The in-toto-verify command basically checks whether the layout hasn't expried and if the steps were performed according to the layout definitions.
+
```
Comment (lukpueh) [inline README.md:12]: IMHO the demo does not need to be explained here, I think the demo README already gives enough context. Just linking to the demo should be fine.

File: README.md [hunk 8]
```diff
@@ -1,2 +1,49 @@
 # in-toto
 in-toto is a series of scripts to protect the supply chain integrity.
+
+# About in-toto
+in-toto guarantees that the end-user (or owner) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (participants) have perfomed the specified tasks and there haven't been any malicious changes in the files.
```
Comment (lukpueh) [inline README.md]: I'd rather not call the end-user "owner", since we are already referring to the actor who writes and signs an in-toto layout as "(project) owner". Also, instead of participants you I'd prefer "(e.g. developers)" or "(carrying out the steps of a software supply chain)".

File: README.md [hunk 9]
```diff
@@ -1,2 +1,49 @@
...
+
+- Alice speficies the layout with Bob and Carl's public keys and signs it with her private key.
+- Bob writes a python script and in-toto hashes the product and stores it to a piece of link metadata. The metadata is signed with Bob's private key. Bob sends the python script to Carl for packaging.
+- Carl packages the script into a tarball and in-toto stores the hash to the metadata.
+- The client (end-user) can now verify the final product which contains the layout, tarball, the both link metadata. The in-toto-verify command basically checks whether the layout hasn't expried and if the steps were performed according to the layout definitions.
+
+# Using in-toto
+* Clone our [repository](https://github.com/in-toto/in-toto.git).
+* Run setup.py
+* Start using the in-toto run and verify scripts
```
Comment (lukpueh) [inline README.md]: Maybe you could call this section something like "Getting started" with the following sub sections:

1. **Installation**
similar to what you wrote in `# Contributing to in-toto`
2. **Create layout**
This will be rather complex. Maybe you can start with a simple description of the parts a layout must/should contain (steps, inspections, functionary keys) and the basic matchrule grammar?
3. **Perform software supply chain steps**
A brief description of the command line tools, `in-toto-run` and `in-toto-record`. What they do and how to use them.
4. **Release final product**
We do not strictly specify how in-toto metadata files should be shipped out together with the final product. For now this section should mention something like:
*In order to verify the final product with in-toto, the verifier must have access to the `layout`, the `*.link` files and the project owner's `public key`.*
5. **Verify final product**
A brief description of the command line tool, `in-toto-verify`.

File: README.md [hunk 10]
```diff
@@ -1,2 +1,49 @@
...
+pip install -e .
+
+*Export the envvar required for "simple settings"*
+export SIMPLE_SETTINGS=in_toto.settings
+
+*Install additional requirements that for some good reason are not in the*
+*requirements file*
+pip install pycrypto cryptography
+
+Submit a Pull Request and start contributing to in-toto.
```
Comment (lukpueh) [inline README.md]: Maybe we can leave out this section for now. Can you please move the installation instructions up to `# Using in-toto` (`# Getting started/Insallation`)?
Also, it would be great if you could run the installation on a vanilla linux user machine (e.g. ubuntu) and find out which system packages are required to install `pycrypto` and `cryptography`.

**Review by lukpueh — DISMISSED** [review]
Thanks, @RohanBhirangi for the README. It is a great start and I appreciate that you are working on this. See my improvement suggestions below.

File: README.md [hunk 11]
```diff
@@ -1,2 +1,54 @@
...
+*Export the envvar required for "simple settings"*
+
+export SIMPLE_SETTINGS=in_toto.settings
+
+*Install additional requirements that for some good reason are not in the*
+*requirements file*
+
+pip install pycrypto cryptography
+
+Submit a Pull Request and start contributing to in-toto.
```
Comment (lukpueh) [inline README.md]: Maybe we can leave out this section for now. Can you please move the installation instructions up to `# Using in-toto` (`# Getting started/Installation`)?
Also, it would be great if you could run the installation on a vanilla linux user machine (e.g. ubuntu) and find out which system packages are required to install `pycrypto` and `cryptography`.

**Comment by SantiagoTorres** [comment]
As a general comment: it's really helpful if you use more descriptive commit messages. You can look at the log to see how we are doing them. I know it's a little counter-intuitive or confusing at first, but you'll get better at it with time 👍

File: README.md [hunk 12]
```diff
@@ -1,2 +1,83 @@
 # in-toto
 in-toto is a series of scripts to protect the supply chain integrity.
+
+# About in-toto
+in-toto guarantees that the end-user (or client) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (eg. developers) have performed the specified tasks and there haven't been any malicious changes in the files.
```
Comment (SantiagoTorres) [inline README.md:4]: I think we could avoid the use of layout without explaining what it is. Overall this feels ok though.

File: README.md [hunk 13]
```diff
@@ -1,2 +1,83 @@
...
+
+in-toto requires a project layout that specifies the functionaries and the tasks they are supposed to perform.
```
Comment (SantiagoTorres) [inline README.md:6]: I would rephrase as:
> in-toto let's project owners to define which steps should be performed and who can perform them to derive a final product.

File: README.md [hunk 14]
```diff
@@ -1,2 +1,83 @@
...
+After each functionary performs its task, a link metadata is generated.
```
Comment (SantiagoTorres) [inline README.md]: `a piece of link metadata is generated` or `a link metadata file is generated` or simply `link metadata is generated`

File: README.md [hunk 15]
```diff
@@ -1,2 +1,83 @@
...
+This metadata is used to verify the intermediate and final products with the project layout.
```
Comment (SantiagoTorres) [inline README.md:8]: This feels a little bit ambiguous. What about:

> This metadata file is used to verify that the materials used and the products produced within this task weren't tampered with. The link metadata also contains a signature to ensure that the right functionary performed this task.

Or something along those lines

File: README.md [hunk 16]
```diff
@@ -1,2 +1,83 @@
...
+
+# in-toto demo
+You can try in-toto by running the demo application at [demo application](https://github.com/in-toto/in-toto/tree/develop/demo).
```
Comment (SantiagoTorres) [inline README.md]: This is ok, but maybe we can delete the repeated link:

> You can try in-toto by running the [demo application](https://github.com/in-toto/in-toto/tree/develop/demo).

File: README.md [hunk 17]
```diff
@@ -1,2 +1,83 @@
...
+The demo basically outlines three users viz., Alice (project owner), Bob (functionary) and Carl (functionary) and how in-toto helps to specify a project layout and verify that the layout has been followed in a correct manner.
+
+# Getting Started
+## 1. Installation
+
+*Make sure you have git, python and pip installed on your system*
+*and get in-toto*
+
+git clone -b develop --recursive https://github.com/in-toto/in-toto.git
```
Comment (SantiagoTorres) [inline README.md:23]: I would:

1.- put this in a code block (you can use  \`\`\` to start and end a code block
2.- Now that the develop branch is the default branch, we don't need the `-b develop` switch. We also need to correct this in the demo text.

File: README.md [hunk 18]
```diff
@@ -1,2 +1,83 @@
...
+
+The layout should specify 6 parts:-
+  1. "_type": which defines a layout
+  2. "expires": which sets the expiry date of the layout
+  3. "inspect": which defines the material and product match rules
+  4. "keys": which specifies the public keys of the functionaries
+  5. "signature": which denotes the private key of the owner
+  6. "steps": which describes the steps involved and the functionaries who are authorized to perform them
+
+You can use the create_layout.py script or write your own script to specify a layout for your project.
```
Comment (SantiagoTorres) [inline README.md]: I like this. Maybe we could link to the create_layout.py script in here also.

File: README.md [hunk 19]
```diff
@@ -1,2 +1,83 @@
...
+  This command will verify that
+    * the layout has not expired,
+    * was signed with the owner's private key, 
+    * and that according to the definitions in the layout
+      * each step was performed and signed by the authorized functionary
+      * the functionaries used the commands they were supposed to use (vi, tar)
+      * the recorded materials and products align with the matchrules and
+      * the inspection untar finds what it expects.
+
+# Specifications
```
Comment (SantiagoTorres) [inline README.md]: I would make this singular.

**Review by SantiagoTorres — CHANGES_REQUESTED** [review]
Some more small notes. It's getting there!

**Review by SantiagoTorres — APPROVED** [review]

**Comment by lukpueh** [comment]
Thanks @RohanBhirangi, for your solid groundwork! I permitted myself to take it a little further before we merge.

File: README.md [hunk 20]
```diff
@@ -1,2 +1,106 @@
 # in-toto
-in-toto is a series of scripts to protect the supply chain integrity.
+Protecting the software supply chain integrity
+
+in-toto guarantees that the end-user (or client) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (eg. developers) have performed the specified tasks and there haven't been any malicious changes in the files.
```
Comment (SantiagoTorres) [inline README.md:4]: I don't know if we should assume that the term "layout" is self-explanatory. What do you guys think?

File: README.md [hunk 21]
```diff
@@ -1,2 +1,106 @@
...
+in-toto requires a project layout that specifies the functionaries and the tasks they are supposed to perform.
+After each functionary performs his or her task a link metadata is generated.
+This metadata is used to verify the intermediate and final products with the project layout.
+
+
+## Getting Started
+
+### Install Dependencies
+ - [Python](www.python.org) in version 2.7 - crypto libraries require header files
+ - [OpenSSL](https://www.openssl.org/) - crypto libraries require header files
```
Comment (SantiagoTorres) [inline README.md:15]: Do we require openssl? who uses openssl as a backend?

Comment (SantiagoTorres) [reply README.md:15]: should we instead mention openssl-dev and stuff like that?

File: README.md [hunk 22]
```diff
@@ -1,2 +1,106 @@
...
+ - [git](https://git-scm.com/) - version control system
+ - [pip](https://pip.pypa.io) - package installer tool
+ - [virtualenvs](http://docs.python-guide.org/en/latest/dev/virtualenvs/) - optional but strongly recommended
```
Comment (SantiagoTorres) [inline README.md:18]: didn't we have a guide on venv + venvwrapper?

File: README.md [hunk 23]
```diff
@@ -1,2 +1,106 @@
...
+### Installation
+```shell
+# Fetch in-toto sources
+git clone -b develop --recursive https://github.com/in-toto/in-toto.git
+
+# Change into project root directory
+cd in-toto
+
+# Install with pip in "develop mode"
+pip install -e .
```
Comment (SantiagoTorres) [inline README.md:29]: Technicallity, but I think it is called "editable" mode. Maybe we want to expand to "--editable" given that we use "--recursive" above.

File: README.md [hunk 24]
```diff
@@ -1,2 +1,106 @@
...
+##### in-toto-run
+`in-toto-run` executes the passed command and records the path and hash of the passed *materials* - files before command execution - and the *products* - files after command execution and optionally stores them together with the *byproducts* of the command - return value, stdout, stderr - to a link file (`<step-name>.link`), signed with the functionary's key.
+
+```shell
+in-toto-run  --step-name <unique step name>
+             --key <functionary private key path>
+            [--materials <filepath>[ <filepath> ...]]
+            [--products <filepath>[ <filepath> ...]]
+            [--record-byproducts]
+            [--verbose] -- <cmd> [args]
```
Comment (SantiagoTorres) [inline README.md:66]: Maybe we don't need all the flags here? If so, I think we are missing a couple.

**Review by SantiagoTorres — CHANGES_REQUESTED** [review]
Looks good!

**Review by SantiagoTorres — APPROVED** [review]
